### ISSUE TITLE
No log output captured in my Pytest HTML report

### ISSUE THREAD

**Issue by nnvkkumar2017** [pytest-dev/pytest-html#444]
Team, every time if i ran the HTML report in pytest getting below error .No logs are captured in that report.Please help me how to resolve this issue to get captured all the data..I have used below code 

pytest -v -s nopecom_testPackage/test_admin_LoginPage.py --html=report.html
![Capture](https://user-images.githubusercontent.com/78788318/107335868-75688480-6ade-11eb-851a-236f6b1371f4.PNG)

**Comment by kotavi** [comment]
I have the same issue when tests are executed in Gitlab pipeline
the command I'm using 
`python.exe -m pytest -s  tests/step_defs/`
Versions: 
pytest - 6.2.2.
pytest-html - 3.1.1

**Comment by gnikonorov** [comment]
I'm a bit tied up with work lately but perhaps @BeyondEvil or @ssbarnea could assist if they have time

**Comment by fenchu** [comment]
Before pytest-html v3 the -s seem to have been ignored/overrided (It actually means do not capture), so it works as intended.

You can use --capture=tee-sys
If it is too much work changing use pytest-html v2.11

**Comment by kotavi** [comment]
@fenchu --capture=tee-sys helped, thank you!

**Comment by shariquealam** [comment]
@fenchu --capture=tee-sys helped, thank you!

**Comment by namigm** [comment]
Thank you, i just removed  -s and it helps me.

**Comment by kisscelia** [comment]
Thanks， i removed -s and it's useful.

**Comment by RajuRekadi** [comment]
yes its working fine when i was remove   -s.

**Comment by Sandeepjiqvia** [comment]
while copying or moving the report.htm to other folder the html styling gets broken every time

**Comment by Sandeepjiqvia** [comment]
command used was pytest -n 5 html=report.html

**Comment by BeyondEvil** [comment]
> command used was pytest -n 5 html=report.html

Yes, you have to also move the assets folder.

Or run with ”—self-contained-html”

**Comment by lefty1975** [comment]
I'm having issues where I'm not capturing log output details for any of my test runs, failed, passed, error, etc. Is this still an open issue or has it been resolved in a newer version?

I'm currently running 3.1.1

**Comment by BeyondEvil** [comment]
@lefty1975 

It's been resolved, but not yet released. I'm working on the final bits as we speak and I'm hoping to release the new version within a week.

Thanks for your patience!

**Comment by lefty1975** [comment]
@BeyondEvil stay evil and thank you!

I thought it was me actually. I did not encounter this issue with other projects using the same version on a mac os. It wasn't until I started a new project in a windows environment that this occurred.

**Comment by kleinyonatan** [comment]
Hi, just got the same issue on pytest-html==3.2.0, was this fixed?

**Comment by BeyondEvil** [comment]
Yeah, sorry no. It wasn’t part of the v3 release, it will be part of the v4 one. Which we are currently wrapping up for a RC release.

**Comment by danielbraun89** [comment]
Log output seems like a core feature that is currently broken, fixing this might make sense as bug fix to v3  instead of v4

**Comment by lefty1975** [comment]
@BeyondEvil I installed the version specified and ran one of my pytests and the report is not functional. Attached is a compressed file of the end result. Maybe this is something on my end that now needs to be adjusted.

[reports.zip](https://github.com/pytest-dev/pytest-html/files/10910566/reports.zip)

**Comment by BeyondEvil** [comment]
> @BeyondEvil I installed the version specified and ran one of my pytests and the report is not functional. Attached is a compressed file of the end result. Maybe this is something on my end that now needs to be adjusted.
> 
> [reports.zip](https://github.com/pytest-dev/pytest-html/files/10910566/reports.zip)

Maybe it's obvious once I open the report.

But can you specify what you mean with "not functional"?

Could you also detail how you generate the report? What flags you're passing in, what OS you're on, versions of python, pytest etc.

Thanks 🙏

**Comment by lefty1975** [comment]
My apologies, that was poor reporting back on my part. 

I will provide clearer details soon.

**Comment by BeyondEvil** [comment]
Can you check if the browser console has any warnings/errors? (right click -> inspect or something like that)

**Comment by BeyondEvil** [comment]
Can you also try with only the `--html=report.html` flag and then add the `--self-contained-html` one?

**Comment by lefty1975** [comment]
I will check for any warnings/errors and run with the flags as outlined.

**Comment by lefty1975** [comment]
@BeyondEvil I believe I have found the issue and it has to do with `https://pypi.org/project/pytest-xdist/` and including the `-n` flag. When I leave this flag out the results are as expected. When adding this flag I receive the previously reported behavior and the Console indicates the following error:
`Uncaught TypeError: data.tests.map is not a function
    at DataManager.setManager (03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:394:55)
    at init (03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:622:13)
    at 4../datamanager.js (03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:629:1)
    at o (03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:388:271)
    at r (03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:388:437)
    at 03_08_2023_09_23_58.html?_ijt=d3ibr5mqhh258krspnji0vd17m&_ij_reload=RELOAD_ON_SAVE:388:466`
    
  I'm also running the latest version of `pytest-xdist` so maybe this is an issue with that library.

**Comment by ssbarnea** [comment]
pytest-xdist is one of the most popular pytest plugins, one that we cannot really ignore. I am using it in ~9/10 projects, very few exceptions and those were basically those where it did not make sense due to concurrency constraints.

**Comment by screig** [comment]
any update?

**Comment by BeyondEvil** [comment]
> any update?

Yes, unless anything unexpected happens I'm releasing 4.0.0rc1 today. It will include a fix for this (and other things). 👍

**Comment by skeng99** [comment]
It works for me when i remove -s .Thanks

**Comment by Anujps** [comment]
> Before pytest-html v3 the -s seem to have been ignored/overrided (It actually means do not capture), so it works as intended.
> 
> You can use --capture=tee-sys If it is too much work changing use pytest-html v2.11

would this work? 
pytest -v -ra --html=report.html --capture=tee-sys

**Comment by BeyondEvil** [comment]
> > Before pytest-html v3 the -s seem to have been ignored/overrided (It actually means do not capture), so it works as intended.
> > You can use --capture=tee-sys If it is too much work changing use pytest-html v2.11
> 
> would this work, i want to see all the print() log on jenkins console and pytest html report. pytest -v -s -ra --html=report.html --capture=tee-sys

`-s` is short for `--capture=no`

```
  --capture=method      Per-test capturing method: one of fd|sys|no|tee-sys
  -s                    Shortcut for --capture=no
```

**Comment by ArcViper** [comment]
I have the same issue as the user who posted the information above. No extra info is showing up in my HTML report. Here is the info I pass in:

py.test --browser_name "%browser%" --html=$WORKSPACE/Reports/report.html -v --junitxml="result.xml" --capture=tee-sys

I do see depreciation warnings in my build output log:

PytestDeprecationWarning: The hookimpl pytest_runtest_makereport uses old-style configuration options (marks or attributes).
  Please use the pytest.hookimpl(hookwrapper=True) decorator instead
   to configure the hooks.
   See https://docs.pytest.org/en/latest/deprecations.html#configuring-hook-specs-impls-using-markers
    @pytest.mark.hookwrapper

also 

py:126: DeprecationWarning: The 'report.extra' attribute is deprecated and will be removed in a future release, use 'report.extras' instead.
    warnings.warn(

**Comment by BeyondEvil** [comment]
@ArcViper What version are you using?

**Comment by markdbuck** [comment]
I am seeing what looks like this same issue with version 4.1.1. Report output is essentially just the following:

There are still tests running. 
Reload this page to get the latest results!

My command invocation is as follows:
```
python3 -m pytest test.py \
  --url="https://${testing_url}/" \
  --username="$user_name" \
  --firstname="$first_name" \
  --password="$password" \
  --userid="$user_id" \
  --self-contained-html \
  --html="WebTesting-${testing_url}.html" \
  --continue-on-collection-errors

```
Here are the loaded plugins:

```
$ pip3 list
Package              Version
-------------------- -----------
attrs                23.2.0
certifi              2023.11.17
charset-normalizer   3.3.2
dnspython            2.4.2
essential-generators 1.0
exceptiongroup       1.2.0
h11                  0.14.0
idna                 3.6
iniconfig            2.0.0
Jinja2               3.1.2
MarkupSafe           2.1.3
outcome              1.3.0.post0
packaging            23.2
pip                  19.3.1
pluggy               1.3.0
PySocks              1.7.1
pytest               7.4.4
pytest-html          4.1.1
pytest-metadata      3.0.0
python-dotenv        1.0.0
requests             2.31.0
selenium             4.16.0
setuptools           41.6.0
smtp-server          0.0.2
sniffio              1.3.0
sortedcontainers     2.4.0
tomli                2.0.1
trio                 0.23.2
trio-websocket       0.11.1
urllib3              2.1.0
webdriver-manager    4.0.1
wsproto              1.2.0
```

**Comment by BeyondEvil** [comment]
Also, could you share a screenshot of what you're seeing and what you're expecting to see?

There are so many things discussed in this issue. 😓

**Comment by markdbuck** [comment]
Here is a screenshot of the html file. As you can see, the main content says "There are still tests running.  Reload this page to get the latest results!".  There are no .ini or .cfg files. Everything works fine with pytest-html version 3.2.0.

![image](https://github.com/pytest-dev/pytest-html/assets/490972/3148b06a-dfad-49a4-b60a-7762c6c50276)

**Comment by BeyondEvil** [comment]
That is really weird. We changed the default behaviour to only generate the report once the suite is finished: https://github.com/pytest-dev/pytest-html/blob/master/src/pytest_html/plugin.py#L76

And since you don't have that flag, you shouldn't see that "button".

What happens if you click it?

**Comment by markdbuck** [comment]
I can't actually click the button. What we are doing is executing a script that runs the tests which creates the html file. Then the next step in the script is to email the html file to a set of recipients. The screen shot above is actually the contents of the email. Is the process of generating the report synchronous?

**Comment by BeyondEvil** [comment]
Sorry for the delay, I'll try and see if I can reproduce this locally. 

Note that setting `generate_report_on_test` to True, will generate the report for each test and _not_ wait until the suite is finished - this gives the "button" behaviour.

**Comment by markdbuck** [comment]
@BeyondEvil thanks for responding and trying to duplicate the issue. I generally don't have generate_report_on_test set to true. I tried it as an experiment but I think I was misunderstanding. Even though the default is supposed to be false maybe I should explicitly set it to false to be sure?

**Comment by BeyondEvil** [comment]
> @BeyondEvil thanks for responding and trying to duplicate the issue. I generally don't have generate_report_on_test set to true. I tried it as an experiment but I think I was misunderstanding. Even though the default is supposed to be false maybe I should explicitly set it to false to be sure?

You shouldn't have to as the [default](https://github.com/pytest-dev/pytest-html/blob/master/src/pytest_html/plugin.py#L78) is "False".

But in the spirit of troubleshooting, try it out and see what happens.

**Comment by markdbuck** [comment]
@BeyondEvil I have more information to share after some additional debugging. The report is, in fact, getting generated at the end of the run. I can open the HTML file in a browser and it looks correct as can be seen in the screenshot below. The problem is that we are sending an email with the html file as the MIME content. When that email is opened in Outlook or Gmail, the HTML does not render correctly and we see the "there are still tests running" button. This can be seen in the second screenshot below. Do you have any thoughts about why the HTML doesn't render in an email? Perhaps you can try this yourself locally. Our python code to send the email with the HTML file as the content is as follows:

```
     msg = MIMEText(open('/app01/web_testing/TestCases/Demo_Site_Tests/TestCases_Filer/JEFS-WebTesting-' + url + '.html').read(), 'html')
    msg['From'] = email_from
    msg['To'] = contacts
    try:
        with smtplib.SMTP('smtp.acme.com', 25) as smtp:
            smtp.send_message(msg)
        print("Email notification sent successfully")
    except Exception as e:
        print(f"Failed to send email notification: (e)")
```

![image](https://github.com/pytest-dev/pytest-html/assets/490972/31226bc6-bef2-40b8-86d2-c8f21345b132)

![image](https://github.com/pytest-dev/pytest-html/assets/490972/8b766d61-872a-4dc7-95f9-9e99ea648225)

**Comment by markdbuck** [comment]
My efforts have been to convert these Selenium tests to run in Kubernetes/Docker, so I didn't look that carefully at the email results with pytest-html 3.2 other than to see that console output was there in the results and when I upgraded to pytest-htm 4.1.1 I saw that the button was there in lieu of any results. But, now that I am looking more carefully, I see that the sorting, expanding, collapsing is NOT working with 3.2 in EMAIL.  Do you have further suggestions?

**Comment by BeyondEvil** [comment]
> My efforts have been to convert these Selenium tests to run in Kubernetes/Docker, so I didn't look that carefully at the email results with pytest-html 3.2 other than to see that console output was there in the results and when I upgraded to pytest-htm 4.1.1 I saw that the button was there in lieu of any results. But, now that I am looking more carefully, I see that the sorting, expanding, collapsing is NOT working with 3.2 in EMAIL. Do you have further suggestions?

Yeah, that confirms my suspicion.

The sorting etc. requires JS, which email editors probably disallows by default for security reasons.

And since 4.x has been refactored to almost entirely be "client side"-rendered, I don't have any quick workarounds unfortunately.

Maybe it's possible to white-list and allow JS in the email - but that isn't a solution that really scales.

Another, also not optimal workaround, would be to what we do in the test suite - basically parse the report via Selenium and then consume it using Beautiful Soup.

Since this has been a recurring theme in issues reported against 4.x, I think a proper solution would be to "pre-render" the report somehow, so that at least the report is "complete", without any rendering needed.

That solution will take some time tho.

**Comment by markdbuck** [comment]
sorry for the delayed response, i have been out sick for a couple of days. understanding that the pre-rendering solution will take a while, i am thinking that the simplest/easiest solution for my org might be to just attach the report to the email rather than embedding it as HTML since the email client won't do JS.

**Comment by BeyondEvil** [comment]
> sorry for the delayed response, i have been out sick for a couple of days. understanding that the pre-rendering solution will take a while, i am thinking that the simplest/easiest solution for my org might be to just attach the report to the email rather than embedding it as HTML since the email client won't do JS.

No worries, hope you're feeling better!

Yes, that should work - because if they then open the attachment, it should open in a browser and be rendered properly.

Let me know the outcome.

**Comment by markdbuck** [comment]
@BeyondEvil - I discussed this with my manager and we are going to change to sending the report as an attachment to the email and just including a summary of the results in the email itself. I verified that when I open the attached report in the email client, it renders correctly in a browser. Thanks again.

**Comment by Pandocstringy** [comment]
@lefty1975 https://pythoner.space/questions/selenium-is-unable-to-access-the-webpage-source-code-when-beautifulsoup-is-used
